LFS-343: Improve the usability of the quick search bar

veronikaslc opened the pull request:
Issues fixed within the LFS-343 PR:
LFS-263: Expand the quick search bar on focus
LFS-327: Clicking on the search result should take user to the highlighted field in the open document
LFS-253: Inconsistent identifiers displayed for forms in the livetable and on the form's page

Commit:
LFS-263: Expand the quick search bar on focus
Used more space for guggestion list elements

Commit:
LFS-253: Inconsistent identifiers displayed for forms in the livetable and on the form's page

Commit:
LFS-327: Clicking on the search result should take user to the highlighted field in the open document

sashaandjic commented:
The search results are now shown in a wider space
![image](https://user-images.githubusercontent.com/23243244/77957887-35235d80-72a2-11ea-87b3-29ce51cfa8f1.png)

sashaandjic commented:
Only one small issue found. If you go to http://172.20.4.87:8080/content.html/Forms/8c701195-6f94-4e33-b6eb-22d6aae3b03e#Donornumbe and type 12 in the search

sashaandjic reviewed (approved):
Tested. The one small issue that I reported is not a show-stopper.

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/headerStyle.jsx:
```diff
@@ -100,7 +100,7 @@ const headerStyle = theme => ({
     whiteSpace: "normal"
   },
   suggestions: {
-    width: theme.spacing(32)
```
The whole search bar should be expanded, not just the suggestions (as shown in the example from github). Alternatively we can make the search bar wider from the beginning. Making just the suggestions wider looks messy to me.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -160,7 +160,7 @@ function Form (props) {
           {
             data && data.subject && data.subject.identifier ?
               <Typography variant="h2">{data.subject.identifier}</Typography>
-            : <Typography variant="h2">{id}</Typography>
```
You could use `Typography`'s `component` and `variant` attributes to ensure the same markup for all titles but smaller font.

However, going forward we won't display these long id, so I wouldn't make this change at all.

Commit:
LFS-253: Inconsistent identifiers displayed for forms in the livetable and on the form's page

Commit:
LFS-263: Expand the quick search bar on focus

sashaandjic reviewed (approved):
tested

OrdiNeu commented on modules/homepage/src/main/frontend/src/themePage/Navbars/SearchBar.jsx:
```diff
@@ -251,9 +252,10 @@ function SearchBar(props) {
                         // Redirect using React-router
                         if (result["@path"]) {
-                          props.history.push("/content.html" + result["@path"]);
+                          props.history.push("/content.html" + result["@path"] + "#" + anchor);
```
Should use the hash argument:
```suggestion
                          props.history.push({
                          pathname: "/content.html" + result["@path"],
                          hash: anchor
                          });
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/FormEntry.jsx:
```diff
@@ -109,12 +47,59 @@ let displaySection = (sectionDefinition, path, depth, existingAnswer, key) => {
-}
+};
+
+export default withStyles(QuestionnaireStyle)(FormEntry);
```
Missing trailing newline.

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/FormEntry.jsx:
```diff
@@ -109,12 +47,59 @@ let displaySection = (sectionDefinition, path, depth, existingAnswer, key) => {
+    );
+  }
+
+  const entry = /#(.+)/.exec(location.href);
```
```suggestion
  const entry = location.hash.substr(1);
```

OrdiNeu reviewed (changes requested):
These changes break conditional Sections
![image](https://user-images.githubusercontent.com/4656440/80031346-8bf21080-84b7-11ea-822f-9b32d985ed9c.png)

Commit:
LFS-343: Improve the usability of the quick search bar
Addressed codereview comments

Commit:
LFS-343: Improve the usability of the quick search bar
Addressing codereview comments

sashaandjic reviewed (approved):
retested

OrdiNeu commented on modules/homepage/src/main/frontend/src/themePage/Navbars/SearchBar.jsx:
```diff
@@ -120,12 +120,13 @@ function SearchBar(props) {
           {(resultData.questionnaire?.title?.concat(' ') || '') + (resultData["jcr:primaryType"]?.replace(/lfs:/,"") || '')}
         </Typography>
-        {resultData.subject?.identifier || resultData["jcr:uuid"] || ''}
+        {resultData.subject?.identifier || entry ? entry[1] : ''}
```
```suggestion
        {resultData.subject?.identifier || entry?.[1] : ''}
```

OrdiNeu reviewed (approved):
Minor note: if you:
1. Search for an answer
2. Select the answer, which brings correctly scrolls you to the answer
3. On the same page, search for the same answer
4. Select the answer again, this time the form will not scroll.

veronikaslc committed:
LFS-343: Improve the usability of the quick search bar

Addressing codereview comments for not scrolling to the section of searched being in the same page.

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/SearchBar.jsx:
```diff
@@ -120,12 +120,13 @@ function SearchBar(props) {
           {(resultData.questionnaire?.title?.concat(' ') || '') + (resultData["jcr:primaryType"]?.replace(/lfs:/,"") || '')}
         </Typography>
-        {resultData.subject?.identifier || resultData["jcr:uuid"] || ''}
+        {resultData.subject?.identifier || entry?.[1] : ''}
```
```suggestion
        {resultData.subject?.identifier || entry?.[1] || ''}
```

veronikaslc committed:
LFS-343: Improve the usability of the quick search bar

Addressing codereview comments

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

sashaandjic reviewed (changes requested):
This subtask no longer works:
 LFS-327: Clicking on the search result should take user to the highlighted field in the open document

VM running on http://172.20.4.131:8080/

Also (not sure if this is withing the scope of this task) some options are showing IDs in the search rather than proper description.  In the example below "Benign" should be shown instead of "B"
![image](https://user-images.githubusercontent.com/23243244/81138360-c8018880-8f2f-11ea-9c25-da7697b4278c.png)

In this example it works as expected: 
![image](https://user-images.githubusercontent.com/23243244/81138418-04cd7f80-8f30-11ea-8f24-9d495effa074.png)

sdumitriu commented on modules/homepage/src/main/frontend/src/themePage/Navbars/headerStyle.jsx:
```diff
@@ -109,7 +110,8 @@ const headerStyle = theme => ({
   aboveBackground: {
-    zIndex: "1301"
+    zIndex: "1301",
+    left: -theme.spacing(3) + "!important"
```
This is invalid since it doesn't have a unit. Didn't test, but this should work:
```suggestion
    left: -theme.spacing(3) + "px !important"
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/LiveTable.jsx:
```diff
@@ -180,6 +180,14 @@ function LiveTable(props) {
 
   let getNestedValue = (entry, path) => {
+    // Display the JCR node id
+    if (path == 'jcr:uuid') {
```
This is not semantically correct, and as a custom code path, it is a code smell. The right approach would be to:
* Add a `@name` value in the JSON ([here](https://github.com/ccmbioinfo/lfs/blob/b6e9d5f454c6f183ba1af1dd2e42dd866dfa78f6/modules/utils/src/main/java/ca/sickkids/ccm/lfs/commons/internal/ResourceToJsonAdapterFactory.java#L134)
* Change the column definition to use `@name` instead of `jcr:uuid` in all table definitions ([here](https://github.com/ccmbioinfo/lfs/blob/b6e9d5f454c6f183ba1af1dd2e42dd866dfa78f6/modules/data-entry/src/main/frontend/src/dataHomepage/Forms.jsx#L61), [here](https://github.com/ccmbioinfo/lfs/blob/b6e9d5f454c6f183ba1af1dd2e42dd866dfa78f6/modules/data-entry/src/main/frontend/src/dataHomepage/UserDashboard.jsx#L40) and [here](https://github.com/ccmbioinfo/lfs/blob/b408ec811637cff118e0aa39f1f33f1160e23458/modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx#L50))

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/FormEntry.jsx:
```diff
@@ -44,16 +44,30 @@ export const ENTRY_TYPES = QUESTION_TYPES.concat(SECTION_TYPES);
  */
-let displayQuestion = (questionDefinition, path, existingAnswer, key) => {
+let displayQuestion = (questionDefinition, path, existingAnswer, key, classes) => {
+  const entry = location.hash.substr(1);
+  const anchor = entry ? entry[1] : '';
```
This is a string, not an array, and as it is it just ends up being just the second character from the anchor.

```suggestion
  const anchor = location.hash.substr(1);
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/FormEntry.jsx:
```diff
@@ -44,16 +44,30 @@ export const ENTRY_TYPES = QUESTION_TYPES.concat(SECTION_TYPES);
+  }, []);
+
+  // if autofocus is needed and specified in the url
+  const questionText = questionDefinition["text"].replace(/\s/g, '').substring(0, 10);
```
This almost works, but is a very fragile way of identifying the question that matched. It doesn't work if there are multiple questions that have the same label, and it doesn't work well with repeated sections. The right question should be referenced by its unique name or path for safety. This does require some changes in the backend as well:
* Change [`getMatchMetadata`](https://github.com/ccmbioinfo/lfs/blob/b6e9d5f454c6f183ba1af1dd2e42dd866dfa78f6/modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/QueryBuilder.java#L224) and its callers to also include the matching node path
* Change `searchBar.jsx` (which you already changed in this PR) to use this path as the hash
* Change this code here to compare the path instead of the question text

sdumitriu commented on modules/homepage/src/main/frontend/src/themePage/Navbars/SearchBar.jsx:
```diff
@@ -1,4 +1,4 @@
-//
+ //
```
```suggestion
//
```

sdumitriu commented on modules/homepage/src/main/frontend/src/themePage/Navbars/SearchBar.jsx:
```diff
@@ -120,12 +120,13 @@ function SearchBar(props) {
   // * the form's subject name (if available) or the resource's uuid
   function QuickSearchResultHeader(props) {
     const {resultData} = props;
+    const entry = /Forms\/(.+)/.exec(resultData["@path"]);
```
If you add `@name` in the JSON this is no longer needed.
```suggestion
```

sdumitriu commented on modules/homepage/src/main/frontend/src/themePage/Navbars/SearchBar.jsx:
```diff
@@ -120,12 +120,13 @@ function SearchBar(props) {
           {(resultData.questionnaire?.title?.concat(' ') || '') + (resultData["jcr:primaryType"]?.replace(/lfs:/,"") || '')}
         </Typography>
-        {resultData.subject?.identifier || resultData["jcr:uuid"] || ''}
+        {resultData.subject?.identifier || entry?.[1] || ''}
```
With the new `@name` that needs to be added:

```suggestion
        {resultData.subject?.identifier || resultData["@name"] || ''}
```

sdumitriu commented on modules/homepage/src/main/frontend/src/themePage/Navbars/headerStyle.jsx:
```diff
@@ -94,13 +94,14 @@ const headerStyle = theme => ({
   search: {
-    marginTop: theme.spacing(1)
+    marginTop: theme.spacing(1),
+    width: theme.spacing(59)
```
`LFS-263` said that the search box should expand only on focus. Is that doable?

veronikaslc replied:
There was an alternative suggestion from Marta that was implemented instead `Alternatively we can make the search bar wider from the beginning. Making just the suggestions wider looks messy to me.` (See the first resolved comment in this PR)

Commit:
LFS-343: Improve the usability of the quick search bar

Addressed codereview comments

sashaandjic commented:
Small issue found: if the search result is within conditional section, the field will be marked properly, but the form would not scroll down to the field

Steps to reproduce:
- go to http://172.20.4.126:8080/
- type "lala"
- pick any of the offered results
- the proper form will be opened, but you need to scroll down manually to see the field you selected in the search result

veronikaslc committed:
LFS-343: Improve the usability of the quick search bar

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

veronikaslc committed:
LFS-343: Improve the usability of the quick search bar

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

veronikaslc committed:
LFS-343: Improve the usability of the quick search bar

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

veronikaslc committed:
LFS-343: Improve the usability of the quick search bar

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

sashaandjic commented:
Small issue still present: if the search result is within conditional section, the field will be marked properly, but the form would not scroll down to the field

Steps to reproduce:
- go to http://172.20.4.87:8080/
- type "kuk"
- click on the 
![image](https://user-images.githubusercontent.com/23243244/81376382-742ea500-90d1-11ea-9df0-6e9392e928ff.png)
- the form opens, but it does not scroll down to the properly marked field

Since this PR has so many other changes, I would merge it. The issue reported above can be logged separately in the back log.

sashaandjic reviewed (approved):
new VM http://172.20.4.116/

Commit:
LFS-343: Improve the usability of the quick search bar
Addressed codereview comments

sashaandjic commented:
VM http://172.20.4.116:8080/
For some reason search/scrolling does not only in the conditional section of Radiotherapy Data form.
Steps to reproduce:
- go to http://172.20.4.116:8080/
- type 4404
- select the Radiotherapy Data form search result
![image](https://user-images.githubusercontent.com/23243244/81535336-9d506f00-9337-11ea-886e-2658d0590901.png)

I can still see ID displayed instead of labels. Here are 2 samples:
Tumor Data form search 
![image](https://user-images.githubusercontent.com/23243244/81535420-c07b1e80-9337-11ea-8dcc-904250e49c0c.png)

Demographic Data form search
![image](https://user-images.githubusercontent.com/23243244/81535609-1cde3e00-9338-11ea-9e53-1906c83a0a0d.png)

Commit:
LFS-343: Improve the usability of the quick search bar
Added timeout for scolling to the conditional section

veronikaslc commented:
@sashaandjic please log ID displayed instead of labels  issue that you found in LFS-343 as a separate issue

sashaandjic commented:
The scrolling is fixed for the first result within conditional section, however if you try to go to the next result within the same form it doesn't move. Step to reporoduce:
- go to http://172.20.4.116:8080/
- type 44 in the search bar
- select the first result --> it works as expected
- now select the second result --> it does not mark nor scroll to the fiels
![image](https://user-images.githubusercontent.com/23243244/81749226-3a66f100-9479-11ea-8a09-a33afe1513cb.png)

sashaandjic reviewed (approved):
TESTED

veronikaslc commented:
This issue was rediscovered 3 times. 
First it was reported in on Mar 30, discussed on a standup, decided to disregard on Mar 31, then reported again on April 30, then reported again now. We need somehow to read the history before reporting issues in the PR, otherwise it becomes completely useless and we spend hours on testing and reporting stuff again.
